lightning-net/src/lib.rs

```diff
@@ -0,0 +1,1098 @@
+// This file is Copyright its original authors, visible in version control
+// history.
+//
+// This file is licensed under the Apache License, Version 2.0 <LICENSE-APACHE
+// or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! # lightning-net
+//!
+//! A socket handling library for those using rust-lightning without an async
+//! runtime.
+//!
+//! Whereas `lightning-net-tokio` manages reading and writing to peers using
+//! Futures and Tokio tasks, this library uses dedicated blocking threads. While
+//! this does result in a small amount of performance overhead, it allows
+//! rust-lightning to be used on platforms that don't support Tokio or async
+//! Rust.
+//!
+//! The primary entrypoints into this crate are `initiate_outbound()` and
+//! `handle_connection()`. See their individual docs for details.
+//!
+//! ## `std` limitations of EDP
+//!
+//! An additional goal of this crate is to compile to the Fortanix EDP
+//! (`x86_64-fortanix-unknown-sgx`) target. This comes with additional
+//! limitations, however. This crate purposefully avoids the use of:
+//!
+//! - `std::time::Instant::now`
+//! - `std::time::Instant::elapsed`
+//! - `std::time::SystemTime::now`
+//! - `std::time::SystemTime::elapsed`
+//! - `std::thread::sleep`
+//! - `std::thread::sleep_ms`
+//! - `std::thread::park_timeout`
+//! - `std::thread::park_timeout_ms`
+//!
+//! These functions have varying degrees of compatibility with Fortanix EDP.
+//! See the [EDP docs](https://edp.fortanix.com/docs/concepts/rust-std/) for
+//! more information on what Rust features can and cannot be used within SGX.
+
+#![deny(broken_intra_doc_links)]
+#![allow(clippy::type_complexity)]
+#![cfg_attr(docsrs, feature(doc_auto_cfg))]
+
+use core::hash;
+use std::io::{Read, Write};
+use std::net::{IpAddr, Shutdown, SocketAddr, TcpStream};
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+use std::thread::{self, JoinHandle};
+
+use crossbeam_channel::{select, Receiver, Sender, TryRecvError, TrySendError};
```

Comment:
We could use the `std` channels, but the other thing used from `crossbeam-channel` is the `select!` macro which operates on crossbeam channel `recv()`s rather than futures. There does not appear to be a `select!` macro (or equivalent functionality) available from `std`.

One approach could be to remove the `select!` by sending `WriteData` requests and `WriterCommand::Shutdown` commands in the same channel, but this introduces a possible race condition due to the need to maintain a bounded `WriteData` queue. What happens is that instead of having a bounded channel of size 1 that only holds `WriteData` requests, we need to have it be of size 2 so that we can always leave a space open for a `Shutdown` command. But in order to leave that second space open, `send_data()` first needs to check `writer_cmd_tx.is_empty()` before actually pushing a `WriteData<Vec<u8>>` into the channel. This could create a TOCTTOU race condition where something else could push into the `writer_cmd` channel in between the `is_empty()` check and the `try_send()` that pushes the write request. Which in turn means that there might not be space for a `Shutdown` command to enter the channel when pushed, which means that a `Reader` or `Writer` thread might be left hanging.

Having only one channel to the Writer was how I implemented the writer commands previously, before splitting it out into two channels to remove the race condition and reduce complexity. [Here's the commit](https://github.com/lexe-tech/rust-lightning/pull/1/commits/6462dbc54a788969bc69141b492d88202dfa4ecb) where I did that, and [here's](https://github.com/lexe-tech/rust-lightning/pull/1/commits/6462dbc54a788969bc69141b492d88202dfa4ecb#diff-6062c086b39587c4e593eeec4310374348ec5082cf9f3bc456ce1f4da53b834fL359) the `is_empty()` check I was referring to.

Reply:
Hmm, I suppose the `writer_cmd` channel capacity could still just be increased to 3. I think the current design is much cleaner and simpler (crossbeam is also more performant), but I'll let you decide.

Reply:
My understanding is that the [`std::mpsc` channels are unmaintained and buggy](https://github.com/rust-lang/rust/issues/39364) : (

My hope is that https://github.com/rust-lang/rust/pull/93563 will go through and just replace the current `std::mpsc` with `crossbeam-channel`. In the meantime, it feels safer to just depend on `crossbeam-channel`.